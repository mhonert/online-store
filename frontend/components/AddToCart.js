import React from 'react';
import Mutation from 'react-apollo/Mutation';
import gql from 'graphql-tag';
import { CURRENT_USER_QUERY } from './User';

export const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART($id: ID!) {
    addToCart(id: $id) {
      quantity
    }
  }
`;

const AddToCart = ({ id }) => {
  return (
    <Mutation
      mutation={ADD_TO_CART_MUTATION}
      variables={{ id }}
      refetchQueries={[{ query: CURRENT_USER_QUERY }]}
    >
      {(addToCart, { loading }) => (
        <button onClick={addToCart} disabled={loading} >Add to Cart</button>
      )}
    </Mutation>
  );
};

export default AddToCart;
