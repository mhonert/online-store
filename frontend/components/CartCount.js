import React from 'react';
import styled from 'styled-components';
import { TransitionGroup, CSSTransition } from 'react-transition-group';

const AnimationStyles = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: all 0.4s;
    backface-visibility: hidden;
  }
  
  .count-enter {
    transform: scale(0.5) rotateX(0.5turn);
  }
  
  .count-enter-active {
    transform: rotateX(0)
  }
  
  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }
  
  .count-exit-active {
    transform: scale(0.5) rotateX(0.5turn)
  }
  
`;

const Dot = styled.div`
  background: ${props => props.theme.primary};
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-weight: 100;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;

const CartCount = ({ cart }) => {
  const count = countItems(cart);
  return (
    <AnimationStyles>
      <TransitionGroup>
        <CSSTransition
          unmountOnExit
          className="count"
          classNames="count"
          key={count}
          timeout={{ enter: 400, exit: 400 }}
        >
          <Dot>{count}</Dot>
        </CSSTransition>
      </TransitionGroup>
    </AnimationStyles>
  );
};

export const countItems = cart =>
  cart.reduce((count, cartItem) => count + cartItem.quantity, 0);

export default CartCount;
