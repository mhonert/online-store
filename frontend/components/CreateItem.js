import React, {useState} from 'react';
import gql from 'graphql-tag';
import { Mutation } from 'react-apollo';
import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import Error from './ErrorMessage';
import Router from 'next/router';

export const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION(
    $title: String!,
    $description: String!,
    $price: Int!,
    $image: String,
    $largeImage: String
    ) {
        createItem(
            title: $title
            description: $description
            price: $price
            image: $image
            largeImage: $largeImage
        ) {
            id
        }
    }
`

const CreateItem = () => {
  const [title, setTitle] = useState('');
  const [description, setDescription] = useState('');
  const [image, setImage] = useState('');
  const [largeImage, setLargeImage] = useState('');
  const [price, setPrice] = useState(0);

  const handleSubmit = async (e, createItem) => {
    // Stop the for from submitting
    e.preventDefault();

    // call the mutation
    const result = await createItem();

    // change them to the single item page
    Router.push({
      pathname: '/item',
      query: {id: result.data.createItem.id}
    });
  }

  return (
    <Mutation mutation={CREATE_ITEM_MUTATION} variables={{title, description, image, largeImage, price}}>
      {(createItem, { loading, error }) => (
        <Form onSubmit={e => handleSubmit(e, createItem)}>
          <Error error={error} />
          <fieldset disabled={loading} aria-busy={loading}>
            <label>Title
              <input type="text" name="title"
                     placeholder="Title" required
                     value={title} onChange={e => setTitle(e.target.value)} />
            </label>
            <label>Price
              <input type="number" name="price"
                     placeholder="Price" required
                     value={price} onChange={e => setPrice(e.target.value)} />
            </label>
            <label>Description
              <textarea type="text" name="description"
                        placeholder="Enter a Description" required
                        value={description} onChange={e => setDescription(e.target.value)} />
            </label>
            <button type="submit">Submit</button>
          </fieldset>
        </Form>
      )}
    </Mutation>
  );
};

export default CreateItem;
