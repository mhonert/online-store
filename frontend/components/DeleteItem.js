import React from 'react';
import gql from 'graphql-tag';
import {Mutation, Query} from 'react-apollo';

export const DELETE_ITEM_MUTATION = gql`
    mutation DELETE_ITEM_MUTATION($id: ID!) {
        deleteItem(id: $id) {
            id
        }
    }
`

const DeleteItem = ({id, onDelete, children}) => {
  const handleDelete = async (e, deleteItemMutation) => {
    if (!confirm("Would you really like to delete this item?")) {
      return;
    }

    const result = await deleteItemMutation().catch(e => alert(e));
  }

  return (
    <Mutation mutation={DELETE_ITEM_MUTATION} variables={{id}} update={onDelete}>
      {(deleteItem, { loading, error }) => (
        <button onClick={e => handleDelete(e, deleteItem)}>{children}</button>
      )}
    </Mutation>
  );
};

export default DeleteItem;
