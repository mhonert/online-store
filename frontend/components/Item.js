import React from 'react';
import PropTypes from 'prop-types';
import ItemStyles from "./styles/ItemStyles";
import Title from "./styles/Title";
import Link from 'next/link';
import PriceTag from "./styles/PriceTag";
import formatMoney from "../lib/formatMoney";

const Item = ({item}) => {
  return (
    <ItemStyles>
      {item.image && <img src={item.image} alt={item.title} />}
      <Title>
        <Link href={{
          pathname: '/item',
          query: { id: item.id }
        }}>
          <a>{item.title}</a>
        </Link>
      </Title>
      <PriceTag>{formatMoney(item.price)}</PriceTag>
      <p>{item.description}</p>

      <div className="buttonList">
        <Link href={{
          pathname: "update",
          query: { id: item.id }
        }}><a>Edit ✏️</a></Link>
        <Link href={{
          pathname: "add",
          query: { id: item.id }
        }}><a>Add to Cart</a></Link>
        <Link href={{
          pathname: "delete",
          query: { id: item.id }
        }}><a>Delete</a></Link>
      </div>
    </ItemStyles>
  );
};

Item.propTypes = {
  item: PropTypes.object.isRequired,
};

export default Item;
