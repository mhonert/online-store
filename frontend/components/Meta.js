import React from 'react';
import Head from 'next/head';

const Meta = () => {
  return (
    <Head>
      <meta name="viewport" content="width=device-width, initial-scale=1" />
      <meta charSet="utf-8" />
      <link rel="stylesheet" type="text/css" href="/nprogress.css" />
      <title>Fresh Fruits</title>
    </Head>
  );
};

export default Meta;