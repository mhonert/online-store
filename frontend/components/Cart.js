import React from 'react';
import CartStyles from './styles/CartStyles';
import CloseButton from './styles/CloseButton';
import Supreme from './styles/Supreme';
import Button from './styles/Button';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import User from './User';
import CartItem from './CartItem';
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';
import { countItems } from './CartCount';
import { adopt } from 'react-adopt';

export const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

export const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

// Compose multiple render props into one component
const Composed = adopt({
  user: ({ render }) => <User>{render}</User>,

  toggleCart: ({ render }) => (
    <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>
  ),

  localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>
});

const Cart = () => {
  return (
    <Composed>
      {({ user, toggleCart, localState }) => {
        const me = user.data.me;
        if (!me) return null;
        const itemCount = countItems(me.cart);
        return (
          <CartStyles open={localState.data.cartOpen}>
            <header>
              <CloseButton title="close" onClick={toggleCart}>
                &times;
              </CloseButton>
              <Supreme>My Cart</Supreme>
              <p>
                You have {itemCount} item
                {itemCount === 1 ? '' : 's'} in your cart.
              </p>
            </header>
            <ul>
              {me.cart.map(cartItem => (
                <CartItem key={cartItem.id} cartItem={cartItem} />
              ))}
            </ul>
            <footer>
              <p>{formatMoney(calcTotalPrice(me.cart))}</p>
              <Button>Checkout</Button>
            </footer>
          </CartStyles>
        );
      }}
    </Composed>
  );
};

export default Cart;
