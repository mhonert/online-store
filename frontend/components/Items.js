import React from 'react';
import {Query} from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import Item from './Item';
import Pagination from './Pagination';
import { perPage } from '../config';

const ALL_ITEMS_QUERY = gql`
    query ALL_ITEMS_QUERY($skip: Int, $first: Int) {
        items(skip: $skip, first: $first, orderBy: createdAt_DESC) {
            id
            title
            price
            description
            image
            largeImage
        }
    }
`;

const Center = styled.div`
  text-align: center;
`;

const ItemsList = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  max-width: ${props => props.theme.maxWidth};
  margin: 0 auto;
`

const Items = ({page}) => {
  const removeItemFromCache = (cache, payload) => {
    // Manually update the cache on the client
    const data = cache.readQuery({query: ALL_ITEMS_QUERY});
    data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id)
    cache.writeQuery({query: ALL_ITEMS_QUERY, data});
  }

  const itemStart = (page - 1) * perPage;

  return (
    <Center>
      <Pagination page={page}/>
      <Query query={ALL_ITEMS_QUERY} variables={{skip: itemStart, first: perPage}}>
        {({data, error, loading}) => {
          if (loading) {
            return <p>Loading ....</p>
          }
          if (error) {
            return <p>Error: {error.message}</p>
          }
          return <ItemsList>
            {data.items.map(item => <Item item={item} key={item.id} onDelete={removeItemFromCache}/>)}
          </ItemsList>
        }}
      </Query>
      <Pagination page={page}/>
    </Center>
  );
};

export default Items;