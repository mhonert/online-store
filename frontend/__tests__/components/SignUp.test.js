import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';

import { MockedProvider } from 'react-apollo/test-utils';
import React from 'react';
import { fakeUser } from '../../lib/testUtils';
import { CURRENT_USER_QUERY } from '../../components/User';
import SignUp, { SIGNUP_MUTATION } from '../../components/SignUp';
import Router from 'next/router';

const me = fakeUser();

// mock the router
Router.router = { push: jest.fn() };

const mocks = [
  // signup mock
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        email: me.email,
        name: me.name,
        password: 'xyz'
      }
    },
    result: {
      data: {
        signup: {
          __typename: 'User',
          id: 'abc123',
          email: me.email,
          name: me.name
        }
      }
    }
  },

  // current user query mock
  {
    request: {
      query: CURRENT_USER_QUERY
    },
    result: {
      data: { me }
    }
  }
];

describe('<SignUp/>', () => {
  it('renders and matches snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <SignUp />
      </MockedProvider>
    );

    expect(toJSON(wrapper.find('form'))).toMatchSnapshot();
  });

  it('user is routed to home page after sign up', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SignUp />
      </MockedProvider>
    );

    await wait();
    wrapper.update();

    simulateInput(wrapper, 'name', me.name);
    simulateInput(wrapper, 'email', me.email);
    simulateInput(wrapper, 'password', 'xyz', 'password');

    wrapper.update();
    wrapper.find('form').simulate('submit');
    await wait();
    wrapper.update();

    expect(Router.router.push).toHaveBeenCalledWith({ pathname: '/' });
  });
});

const simulateInput = (wrapper, name, value, type) => {
  wrapper
    .find(`input[name="${name}"]`)
    .simulate('change', { target: { name, value, type } });
};
