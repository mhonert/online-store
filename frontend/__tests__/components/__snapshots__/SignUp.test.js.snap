// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp/> renders and matches snapshot 1`] = `
<form
  className="Form-sc-3ek8g3-0 kxkTjx"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Register An Account
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label>
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="Email"
        required={true}
        type="email"
        value=""
      />
    </label>
    <label>
      Name
      <input
        name="name"
        onChange={[Function]}
        placeholder="Name"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label>
      Password
      <input
        name="password"
        onChange={[Function]}
        placeholder="Password"
        required={true}
        type="password"
        value=""
      />
    </label>
  </fieldset>
  <button
    type="submit"
  >
    Register!
  </button>
</form>
`;
